Sources/Toolbox/Utilities/FailableCodable.swift

```diff
@@ -0,0 +1,56 @@
+//
+//  FailableCodable.swift
+//  
+//
+//  Created by Oliver Krakora on 06.11.20.
+//
+
+import Foundation
+
+/// A helper that allows you to wrap potential faulty `Codable` data structures
+///
+/// **Usage**
+///
+///     struct Person: Codable {
+///         let name: String
+///         let websiteURL: FailableCodable<URL>
+///     }
+///
+/// The `websiteURL` is defined by a user so there is no guarantee that it is a valid `URL`.
+///
+///
+/// Without `FailableCodable` the decoding would fail for the entire `Person` and you would have to implement the decoding intializer yourself.
+///
+enum FailableCodable<T: Codable> {
+    /// Indicates that decoding of the underlying data was successfull
+    case success(T)
+    /// Indicates that the underlying data failed to decode
+    case error(Error)
+    
+    /// If the decoding was successfull a non nil value of type `T`
+    var value: T? {
+        switch self {
+        case .success(let value):
+            return value
+        case .error:
+            return nil
+        }
+    }
+    
+    init(from decoder: Decoder) throws {
+        do {
+            self = try .success(decoder.singleValueContainer().decode(T.self))
+        } catch {
+            self = .error(error)
+        }
+    }
+    
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        if let value = value {
+            try container.encode(value)
+        } else {
+            try container.encodeNil()
```

Comment:
Decoding error case should throw an error. I think it's an implementation detail how to handle error case.

Reply:
hm, I wouldn't throw an error. Imagine you fail to decode some property, resulting into an error case (that's what the whole FailableCodable is for) and then you want to store it, send back to the server, etc. This shouldn't throw an error. So I think the current behaviour is correct.

Reply:
You are right, this is the purpose of FailableCodable ;-)